pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,425 @@
+        self._init_gradients(intg)
+
+        # Base directory for output and counter reset
+        self.basedir = self.cfg.getpath(cfgsect, 'basedir', '.', abs=True)
+        for sname, scfg in self._scene_cfg.items():
+            scfg['path_gen'] = file_path_gen(self.basedir, scfg['image_name'])
+            next(scfg['path_gen'])
+
+        # Set subdivision
+        self.divisor = self.cfg.getint(cfgsect, 'division', 
+                                       self.cfg.getint('solver', 'order'))
+        self.etypes_div = defaultdict(lambda: self.divisor)
+
+        # Share number of ele types per part
+        ndoms = len(system.ele_map)
+        ndoms = comm.gather(ndoms, root=root)
+        if rank == root:
+            ndoms = np.asarray(ndoms)
+        ndoms = comm.bcast(ndoms, root=root)
+        self.dom_0 = dom_0 = np.sum(ndoms[:rank])
+
+        # Generate conduit blueprint of mesh
+        mesh_n = ConduitNode(self.conduit)
+        self.ops = {}
+
+        for idx, ename, rgn in self._ele_regions:
+            if ename in ['pri', 'pyr']:
+                raise ValueError(f'Ascent: {ename} elements not supported')
+            eles = system.ele_map[ename]
+
+            dpf = f'domain_{dom_0 + idx}'
+
+            shape = subclass_where(BaseShape, name=ename)(eles.nspts, 
+                                                          self.cfg)
+
+            svpts = shape.std_ele(self.etypes_div[ename])
+            nsvpts = len(svpts)
+
+            mesh_op = shape.sbasis.nodal_basis_at(svpts).astype(self.dtype)
+            soln_op = shape.ubasis.nodal_basis_at(svpts).astype(self.dtype)
+            self.ops[ename] = (mesh_op, soln_op)
+
+            mesh_n[f'{dpf}/state/domain_id'] = rank
+            mesh_n[f'{dpf}/coordsets/coords/type'] = 'explicit'
+            mesh_n[f'{dpf}/topologies/mesh/coordset'] = 'coords'
+            mesh_n[f'{dpf}/topologies/mesh/type'] = 'unstructured'
+            mesh_n[f'{dpf}/topologies/mesh/elements/shape'] = ename
+
+            xd = np.einsum('ik,k...', mesh_op, 
+                           eles.ploc_at_np('linspts')[..., rgn])
+            for l , x in zip('xyz', xd.reshape(eles.ndims, -1)):
+                mesh_n[f'{dpf}/coordsets/coords/values/{l}'] = x
+
+            subdvcls = subclass_where(BaseShapeSubDiv, name=ename)
+            sconn = subdvcls.subnodes(self.divisor)
+            neles_rgn = xd.shape[1]
+            subdvcon = np.hstack([sconn + j*nsvpts for j in range(neles_rgn)])
+            mesh_n[f'{dpf}/topologies/mesh/elements/connectivity'] = subdvcon
+
+            for field in self._expr:
+                mesh_n[f'{dpf}/fields/{field}/association'] = 'vertex'
+                mesh_n[f'{dpf}/fields/{field}/volume_dependent'] = 'false'
+                mesh_n[f'{dpf}/fields/{field}/topology'] = 'mesh'
+
+        self.mesh_n = mesh_n
+
+        # Inital Ascent setup
+        self.lib = AscentWrappers()
+        self._as_parameter_ = c_void_p(self.lib.ascent_create(None))
```

Comment:
Probably shouldn't use _as_parameter_ here; really make no sense that if you were to pass a plugin object to a C function you get the ascent object.  Cast is also not needed.